documentation/puppet-api/v4/catalog.markdown

```diff
@@ -0,0 +1,67 @@
+---
+layout: default
+title: "Puppet Server: Puppet API: Catalog"
+canonical: "/puppetserver/latest/puppet-api/v4/catalog.html"
+---
+
+The catalog API returns a compiled catalog for the node specified in the request,
+making use of provided metadata like facts or environment if specified.
+If not specified, it will attempt to fetch this data from Puppet's configured sources
+(usually PuppetDB or a node classifier). The returned catalog is in JSON format,
+ready to be parsed and applied by an agent.
+
+## `POST /puppet/v4/catalog`
+
+(Introduced in Puppet Server 6.3.0)
+
+The input data for the catalog to be compiled is submitted as a JSON body with the 
+following form:
+```
+{
+  "certname": "<node name>",
+  "persistence": { "facts": <true/false>, "catalog": <true/false> },
+  # The rest are optional:
+  "facts": { "values": { "<fact name>": <fact value>, ... } },
+  "trusted_facts": { "values": { "<fact name>": <fact value>, ... } },
+  "environment": "<environment name>",
+  "prefer_requested_environment": <true/false>
+}
+```
+
+#### `certname` (required)
+The name of the node for which to compile the catalog.
+
+#### `persistence` (required)
+A hash containing two required keys, `facts` and `catalog`, which when set to true will
+cause the facts and reports to be stored in PuppetDB, or discarded if set to false.
+
+#### `facts`
+A hash with a required `values` key, containing a hash of all the facts for the node.
+If not provided, Puppet will attempt to fetch facts for the node from PuppetDB.
+
+#### `trusted_facts`
+A hash with a required `values` key containing a hash of the trusted facts for a node.
+In a normal agent's catalog request, these would be extracted from the cert, but this
+endpoint does not require a cert for the node whose catalog is being compiled. If not
+provided, Puppet will attempt to fetch the trusted facts for the node from PuppetDB or
+from the provided facts hash.
+
+#### `environment`
+The name of the environment for which to compile the catalog. If `prefer_requested_environemnt`
+is true, override the classified environment with this param. If it is false, only respect this
+if the classifier allows an agent-specified environment.
+
+#### `prefer_requested_environment`
```

Comment:
I know, this isn't what we talked about, but can we just say, if the api caller has specified the environment use it, if they haven't specified it use the classifier, and if not classified fall back to production/whatever's default?

Reply:
I'm pretty sure we need to support the normal rules around agent-specified environments for ACE compilations though... I guess we can see if anyone responds to the email I sent that team overnight.

Reply:
@DavidS thoughts on that question? Do agentless nodes obey the same classifier environment rules as normal nodes?

Reply:
Got a "yes" on this from Alex.

Reply:
This is kind of nasty, but I think unfortunately necessary.